src/mlmc/quantity.py

```diff
@@ -318,30 +303,61 @@ def ne_op(x, y):
             return self._process_mask(x, y, operator.ne)
         return self._mask_quantity(other, ne_op)
 
-    def subsample(self, sample_vec):
-        """
-        Random subsampling
-        :param sample_vec: list of number of samples at each level
+    @staticmethod
+    def pick_samples(chunk, subsample_params):
+        """
+        Pick samples some samples from chunk in order to have 'k' samples from 'n' after all chunks are processed
+        Note that it is not guaranteed to receive the exact number of 'k' samples
+        :param chunk: np.ndarray, shape M, N, 2, where N denotes number of samples in chunk
+        :param subsample_params: instance of SubsampleParams class, it has two parameters:
+                        k: number of samples which we want to get from all chunks
+                        n: number of all samples among all chunks
         :return: np.ndarray
         """
-        quantity_storage = self.get_quantity_storage()
-        n_collected = quantity_storage.n_collected()
+        out = []
+        for _ in range(chunk.shape[1]):
+            i = random.randint(0, subsample_params.n)
+            if i < subsample_params.k and i < chunk.shape[1]:
+                out.append(chunk[:, i, :])
+                subsample_params.k -= 1
+            subsample_params.n -= 1
```

Comment:
Clever, yet modification of the input quantity violates functional programming paradigm, which may block possible parallelization etc. If we adopt chunk iterator, then this information can be part of it. Still not the ideal solution.